From: Matthew Hagemann <matt.hagemann@canonical.com>
Date: Thu, 13 Feb 2025 17:08:34 +0200
Subject: Dynamic lang pack installation support

Adding a polkit rule to allow the gnome initial setup user to run pkexec if the
program it is running is apt. This is to allow for Gnome-Initial-Setup install
language packs during setup via apt update && apt install.

Adding a function apt_install that takes in a list of packages and attempts to
install them. This is inteded to be used as part of the new flow for
dynamically installing language packs for the user during setup.

Added parse_language_list function to parse the static list of languages in
/usr/share/language-selector/data/languagelist which is now used to populate
the language chooser. Gnome lang packs are used as a fallback.

Added the ability for GIS to determine what outstanding language packs need to
be installed via check-language-support -l and to use that output to install
outstanding lang packs via a g_task.

Added a spinner to the language chooser for when a language's packs are being
installed. All other rows and next button are set to sensitive FALSE while
GIS attempts to install language packs. If language pack installation fails,
a warning is shown to the user and the locale is set. While waiting the user
has the option to cancel the installation of additional lang packs.

Additional PR style updates

PR code styple cleanup
---
 data/20-gnome-initial-setup.rules.in               |  10 +
 gnome-initial-setup/gis-pkexec.c                   |  65 ++++-
 gnome-initial-setup/gis-pkexec.h                   |   6 +
 .../pages/language/cc-language-chooser.c           | 289 ++++++++++++++++++++-
 .../pages/language/cc-language-chooser.h           |   1 +
 .../pages/language/gis-language-page.c             |  20 +-
 .../pages/software/gis-software-page.c             |   2 +-
 7 files changed, 380 insertions(+), 13 deletions(-)

diff --git a/data/20-gnome-initial-setup.rules.in b/data/20-gnome-initial-setup.rules.in
index 02fd21d..d769020 100644
--- a/data/20-gnome-initial-setup.rules.in
+++ b/data/20-gnome-initial-setup.rules.in
@@ -5,6 +5,16 @@
 // Allow the gnome-initial-setup user to do certain actions without
 // being interrupted by password dialogs
 
+polkit.addRule(function(action, subject) {
+    if (subject.user !== 'gnome-initial-setup')
+        return undefined;
+
+    if (action.id === 'org.freedesktop.policykit.exec' &&
+        action.lookup('program') === '/usr/bin/apt-get') {
+        return 'yes';
+    }
+});
+
 polkit.addRule(function(action, subject) {
     if (subject.user !== 'gnome-initial-setup')
         return undefined;
diff --git a/gnome-initial-setup/gis-pkexec.c b/gnome-initial-setup/gis-pkexec.c
index 9e1bfa9..6654b00 100644
--- a/gnome-initial-setup/gis-pkexec.c
+++ b/gnome-initial-setup/gis-pkexec.c
@@ -29,6 +29,7 @@ gboolean
 gis_pkexec (const gchar  *command,
             const gchar  *arg1,
             const gchar  *user,
+            GCancellable *cancellable,
             GError      **error)
 {
   g_autoptr(GSubprocessLauncher) launcher = NULL;
@@ -43,6 +44,8 @@ gis_pkexec (const gchar  *command,
    * so remove it from the environment.
    */
   g_subprocess_launcher_unsetenv (launcher, "SHELL");
+  g_subprocess_launcher_setenv (launcher, "DEBCONF_NONINTERACTIVE_SEEN", "true", TRUE);
+  g_subprocess_launcher_setenv (launcher, "DEBIAN_FRONTEND", "noninteractive", TRUE);
   process = g_subprocess_launcher_spawnv (launcher, argv, error);
 
   if (!process) {
@@ -50,10 +53,70 @@ gis_pkexec (const gchar  *command,
     return FALSE;
   }
 
-  if (!g_subprocess_wait_check (process, NULL, error)) {
+  if (!g_subprocess_wait_check (process, cancellable, error)) {
     g_prefix_error (error, "%s failed: ", command);
     return FALSE;
   }
 
   return TRUE;
 }
+
+gboolean
+apt_get_install (GPtrArray     *packages,
+                 GCancellable  *cancellable,
+                 GError       **error) {
+  g_autoptr(GSubprocessLauncher) launcher = NULL;
+  g_autoptr(GSubprocess) process = NULL;
+  static gboolean already_updated = FALSE;
+
+  g_return_val_if_fail (packages != NULL, FALSE);
+  g_return_val_if_fail (packages->len != 0, FALSE);
+
+  if (!g_atomic_int_get (&already_updated)) {
+    /* Avoid running apt-update again if it has been already completed in a previous iteration */
+    if (!gis_pkexec ("/usr/bin/apt-get", "update", NULL, cancellable, error))
+      return FALSE;
+
+    g_atomic_int_set (&already_updated, TRUE);
+  }
+
+  /* Create space for 8 leading arguments of the install command, the number of
+   * packages and a NULL terminator.
+   */
+  int total = 4 + packages->len + 1;
+  g_autoptr (GPtrArray) argv = g_ptr_array_new_null_terminated (total, g_free, TRUE);
+
+  g_ptr_array_add (argv, g_strdup ("/usr/bin/pkexec"));
+  g_ptr_array_add (argv, g_strdup ("/usr/bin/apt-get"));
+  g_ptr_array_add (argv, g_strdup ("install"));
+  g_ptr_array_add (argv, g_strdup ("-y"));
+  while (packages->len) {
+    g_ptr_array_add (argv, g_ptr_array_steal_index_fast (packages, 0));
+  }
+
+  launcher = g_subprocess_launcher_new (G_SUBPROCESS_FLAGS_NONE);
+  g_subprocess_launcher_unsetenv (launcher, "SHELL");
+
+  g_subprocess_launcher_setenv (launcher, "DEBCONF_NONINTERACTIVE_SEEN", "true", TRUE);
+  g_subprocess_launcher_setenv (launcher, "DEBIAN_FRONTEND", "noninteractive", TRUE);
+
+  process =
+      g_subprocess_launcher_spawnv (launcher, (const gchar *const *) argv->pdata, error);
+
+  if (!process) {
+    g_prefix_error_literal (error, "Failed to create apt install process: ");
+    return FALSE;
+  }
+
+  if (g_cancellable_set_error_if_cancelled (cancellable, error)) {
+    g_prefix_error_literal (error, "User canceled subprocess: ");
+    return FALSE;
+  }
+
+  if (!g_subprocess_wait_check (process, cancellable, error)) {
+    g_prefix_error_literal (error, "apt-get install failed: ");
+    return FALSE;
+  }
+
+  return TRUE;
+}
diff --git a/gnome-initial-setup/gis-pkexec.h b/gnome-initial-setup/gis-pkexec.h
index ff5a88c..000bea9 100644
--- a/gnome-initial-setup/gis-pkexec.h
+++ b/gnome-initial-setup/gis-pkexec.h
@@ -34,8 +34,14 @@ gboolean
 gis_pkexec (const gchar  *command,
             const gchar  *arg1,
             const gchar  *user,
+            GCancellable *cancellable,
             GError      **error);
 
+gboolean
+apt_get_install (GPtrArray     *packages,
+                 GCancellable  *cancellable,
+                 GError       **error);
+
 G_END_DECLS
 
 #endif /* __GIS_PKEXEC_H__ */
diff --git a/gnome-initial-setup/pages/language/cc-language-chooser.c b/gnome-initial-setup/pages/language/cc-language-chooser.c
index 61756be..fd05f17 100644
--- a/gnome-initial-setup/pages/language/cc-language-chooser.c
+++ b/gnome-initial-setup/pages/language/cc-language-chooser.c
@@ -21,6 +21,7 @@
  */
 
 #include "config.h"
+#include "gis-pkexec.h"
 #include "cc-language-chooser.h"
 
 #include <locale.h>
@@ -30,6 +31,8 @@
 #include <gtk/gtk.h>
 
 #define GNOME_DESKTOP_USE_UNSTABLE_API
+#define LANGUAGE_LIST_PATH "/usr/share/language-selector/data/languagelist"
+
 #include <libgnome-desktop/gnome-languages.h>
 
 #include "cc-common-language.h"
@@ -45,8 +48,11 @@ struct _CcLanguageChooserPrivate
         GtkWidget *no_results;
         GtkWidget *more_item;
 
+        gboolean busy;
         gboolean showing_extra;
         gchar *language;
+
+        GCancellable *install_cancellable;
 };
 typedef struct _CcLanguageChooserPrivate CcLanguageChooserPrivate;
 G_DEFINE_TYPE_WITH_PRIVATE (CcLanguageChooser, cc_language_chooser, GTK_TYPE_BOX);
@@ -55,6 +61,7 @@ enum {
         PROP_0,
         PROP_LANGUAGE,
         PROP_SHOWING_EXTRA,
+        PROP_BUSY,
         PROP_LAST,
 };
 
@@ -71,6 +78,11 @@ typedef struct {
         GtkWidget *box;
         GtkWidget *checkmark;
 
+        GtkWidget *status_stack;
+        GtkWidget *spinner;
+        GtkWidget *status_label;
+        GtkWidget *cancel_button;
+
         gchar *locale_id;
         gchar *locale_name;
         gchar *locale_current_name;
@@ -113,9 +125,17 @@ language_widget_free (gpointer data)
         g_free (widget);
 }
 
+static void
+on_cancel_button_clicked (CcLanguageChooser *chooser)
+{
+    CcLanguageChooserPrivate *priv = cc_language_chooser_get_instance_private (chooser);
+    g_cancellable_cancel (priv->install_cancellable);
+}
+
 static GtkWidget *
-language_widget_new (const char *locale_id,
-                     gboolean    is_extra)
+language_widget_new (CcLanguageChooser *chooser,
+                     const char        *locale_id,
+                     gboolean           is_extra)
 {
 	GtkWidget *label;
         gchar *locale_name, *locale_current_name, *locale_untranslated_name;
@@ -155,8 +175,25 @@ language_widget_new (const char *locale_id,
         gtk_label_set_xalign (GTK_LABEL (label), 0);
         gtk_box_append (GTK_BOX (widget->box), label);
 
+        widget->status_stack = gtk_stack_new();
+
         widget->checkmark = gtk_image_new_from_icon_name ("object-select-symbolic");
-        gtk_box_append (GTK_BOX (widget->box), widget->checkmark);
+        gtk_stack_add_titled (GTK_STACK (widget->status_stack), widget->checkmark, "check", "Checkmark");
+
+        widget->spinner = gtk_spinner_new();
+        gtk_stack_add_titled (GTK_STACK (widget->status_stack), widget->spinner, "spinner", "Spinner");
+
+        widget->status_label = gtk_label_new ("");
+        gtk_widget_set_name(widget->status_label, "status");
+        gtk_widget_add_css_class(widget->status_label, "dim-label");
+
+        widget->cancel_button = gtk_button_new_with_label (_("Cancel"));
+        gtk_widget_set_visible (widget->cancel_button, FALSE);
+
+        gtk_stack_set_visible_child_name (GTK_STACK (widget->status_stack), "check");
+        gtk_box_append (GTK_BOX (widget->box), widget->status_stack);
+        gtk_box_append (GTK_BOX (widget->box), widget->cancel_button);
+        gtk_box_append (GTK_BOX (widget->box), widget->status_label);
 
         if (country_name) {
                 label = gtk_label_new (country_name);
@@ -179,6 +216,10 @@ language_widget_new (const char *locale_id,
         g_object_set_data_full (G_OBJECT (widget->box), "language-widget", widget,
                                 language_widget_free);
 
+        g_signal_connect_object (widget->cancel_button, "clicked",
+                                 G_CALLBACK (on_cancel_button_clicked),
+                                 chooser, G_CONNECT_SWAPPED);
+
         g_free (language);
         g_free (language_name);
         g_free (country);
@@ -188,11 +229,57 @@ language_widget_new (const char *locale_id,
 }
 
 static void
-sync_all_checkmarks (CcLanguageChooser *chooser)
+update_language_widget_visibility (LanguageWidget *widget,
+                                   GtkWidget      *row,
+                                   gboolean        waiting,
+                                   gboolean        visible,
+                                   const gchar     *error_message)
+{
+        if (visible) {
+                if (waiting) {
+                        gtk_stack_set_visible_child_name (GTK_STACK (widget->status_stack),
+                                                          "spinner");
+                        gtk_spinner_start (GTK_SPINNER (widget->spinner));
+                        gtk_label_set_text (GTK_LABEL (widget->status_label),
+                                            _("Downloading language packs…"));
+                        gtk_widget_set_visible (widget->cancel_button, TRUE);
+                } else {
+                        gtk_spinner_stop (GTK_SPINNER (widget->spinner));
+                        gtk_stack_set_visible_child_name (GTK_STACK (widget->status_stack),
+                                                          "check");
+                        gtk_widget_set_visible (widget->cancel_button, FALSE);
+
+                        if (error_message) {
+                                gtk_label_set_text (GTK_LABEL (widget->status_label),
+                                                    error_message);
+                        } else {
+                                gtk_label_set_text (GTK_LABEL (widget->status_label), "");
+                        }
+                }
+        } else {
+                gtk_stack_set_visible_child_name (GTK_STACK (widget->status_stack), "check");
+                gtk_label_set_text (GTK_LABEL (widget->status_label), "");
+                gtk_widget_set_visible (widget->cancel_button, FALSE);
+        }
+
+        if (waiting && !visible) {
+                gtk_widget_set_sensitive (GTK_WIDGET (row), FALSE);
+        } else {
+                gtk_widget_set_sensitive (GTK_WIDGET (row), TRUE);
+        }
+}
+
+static void
+sync_all_checkmarks (CcLanguageChooser *chooser,
+                     gboolean           waiting,
+                     const gchar       *error_message)
 {
         CcLanguageChooserPrivate *priv = cc_language_chooser_get_instance_private (chooser);
         GtkWidget *row;
 
+        priv->busy = waiting;
+        g_object_notify_by_pspec (G_OBJECT (chooser), obj_props[PROP_BUSY]);
+
         row = gtk_widget_get_first_child (priv->language_list);
         while (row) {
                 LanguageWidget *widget;
@@ -206,8 +293,8 @@ sync_all_checkmarks (CcLanguageChooser *chooser)
                         return;
 
                 should_be_visible = g_str_equal (widget->locale_id, priv->language);
+                update_language_widget_visibility (widget, row, waiting, should_be_visible, error_message);
                 gtk_widget_set_opacity (widget->checkmark, should_be_visible ? 1.0 : 0.0);
-
                 row = gtk_widget_get_next_sibling (row);
         }
 }
@@ -257,7 +344,7 @@ add_one_language (CcLanguageChooser *chooser,
 		return;
 	}
 
-	widget = language_widget_new (locale_id, !is_initial);
+	widget = language_widget_new (chooser, locale_id, !is_initial);
         if (widget)
                 gtk_list_box_append (GTK_LIST_BOX (priv->language_list), widget);
 }
@@ -290,13 +377,54 @@ add_languages (CcLanguageChooser  *chooser,
         gtk_list_box_set_placeholder (GTK_LIST_BOX (priv->language_list), priv->no_results);
 }
 
+static GStrv
+parse_language_list (void)
+{
+    g_autofree gchar *contents = NULL;
+    g_autoptr (GError) error = NULL;
+
+    if (!g_file_get_contents (LANGUAGE_LIST_PATH, &contents, NULL, &error)) {
+        g_critical ("Error reading file: %s", error->message);
+        return NULL;
+    }
+
+    g_auto (GStrv) lines = g_strsplit (contents, "\n", -1);
+    g_autoptr (GStrvBuilder) builder = g_strv_builder_new ();
+
+    for (gchar **line = lines; *line != NULL; line++) {
+        const gchar *trimmed = g_strstrip (*line);
+
+        /* Skip empty lines and commented out lines */
+        if (trimmed[0] == '\0' || trimmed[0] == '#') {
+            continue;
+        }
+
+        g_auto (GStrv) tokens = g_strsplit (trimmed, ";", -1);
+        gint token_count = g_strv_length (tokens);
+
+        /* Language code we use is the 4th entry in a ; delimited list */
+        if (token_count >= 4)
+            g_strv_builder_add (builder, g_strdup (tokens[3]));
+    }
+
+    return g_strv_builder_end (builder);
+}
+
 static void
 add_all_languages (CcLanguageChooser *chooser)
 {
         g_auto(GStrv) locale_ids = NULL;
         g_autoptr(GHashTable) initial = NULL;
 
-        locale_ids = gnome_get_all_locales ();
+        /* Make use of the static language list in
+         * /usr/share/language-selector/data/languagelist. If not present, fall
+         * back to using gnome_get_all_locales().
+         */
+        if (!(locale_ids = parse_language_list ())) {
+                g_warn_if_reached ();
+                locale_ids = gnome_get_all_locales ();
+        }
+
         initial = cc_common_language_get_initial_languages ();
         add_languages (chooser, locale_ids, initial);
 }
@@ -395,6 +523,119 @@ show_more (CcLanguageChooser *chooser)
         g_object_notify_by_pspec (G_OBJECT (chooser), obj_props[PROP_SHOWING_EXTRA]);
 }
 
+static GPtrArray *
+get_lang_support_packages_for_locale (const gchar   *locale,
+                                      GCancellable  *cancellable,
+                                      GError       **error)
+{
+        g_autoptr (GSubprocess) subprocess = NULL;
+        g_autofree gchar *stdout_buf = NULL;
+        g_autofree gchar *stderr_buf = NULL;
+
+        subprocess = g_subprocess_new (G_SUBPROCESS_FLAGS_STDOUT_PIPE |
+                                       G_SUBPROCESS_FLAGS_STDERR_PIPE,
+                                       error,
+                                       "/usr/bin/check-language-support",
+                                       "-l",
+                                       locale,
+                                       NULL);
+
+        if (!subprocess)
+                return NULL;
+
+        if (!g_subprocess_communicate_utf8 (subprocess, NULL, cancellable,
+                                            &stdout_buf, &stderr_buf, error)) {
+                g_prefix_error_literal (error, "Failed to communicate with check-language-support: ");
+                return NULL;
+        }
+
+        if (!g_subprocess_wait_check (subprocess, cancellable, error)) {
+                g_prefix_error_literal (error, "check-language-support returned an error: ");
+                return NULL;
+        }
+
+        if (stderr_buf && *stderr_buf != '\0')
+                g_warning ("check-language-support stderr: %s", stderr_buf);
+
+        if (!stdout_buf || *stdout_buf == '\0')
+                return g_ptr_array_new_with_free_func (g_free);
+
+        g_auto (GStrv) packages_split = g_strsplit (g_strstrip (stdout_buf), " ", -1);
+        return g_ptr_array_new_take_null_terminated ((gpointer *) g_steal_pointer (&packages_split), g_free);
+}
+
+static gboolean
+install_language_support (const gchar   *locale,
+                          GCancellable  *cancellable,
+                          GError       **error)
+{
+        g_autoptr (GPtrArray) packages = NULL;
+
+        packages = get_lang_support_packages_for_locale (locale, cancellable, error);
+        if (!packages)
+                return FALSE;
+
+        if (packages->len == 0) {
+                g_message ("No packages to install for locale: %s", locale);
+                return TRUE;
+        }
+
+        if (!apt_get_install (packages, cancellable, error)) {
+                g_prefix_error_literal (error, "Error installing language support: ");
+                return FALSE;
+        }
+
+        return TRUE;
+}
+
+static void
+installation_task_async (GTask        *task,
+                         gpointer      source_object,
+                         gpointer      task_data,
+                         GCancellable *cancellable)
+{
+        const gchar *locale = task_data;
+        g_autoptr (GError) error = NULL;
+
+        if (g_cancellable_set_error_if_cancelled (cancellable, &error)) {
+                g_task_return_error (task, g_steal_pointer(&error));
+                return;
+        }
+
+        if (!install_language_support (locale, cancellable, &error)) {
+                g_task_return_error (task, g_steal_pointer (&error));
+                return;
+        }
+
+        g_task_return_boolean(task, TRUE);
+}
+
+static void
+installation_task_done (GObject      *source_object,
+                        GAsyncResult *res,
+                        gpointer      user_data)
+{
+        g_autoptr (GError) error = NULL;
+        CcLanguageChooser *chooser = user_data;
+
+        if (g_task_propagate_boolean (G_TASK (res), &error)) {
+                g_object_notify_by_pspec (G_OBJECT (chooser),
+                                          obj_props[PROP_LANGUAGE]);
+                sync_all_checkmarks (chooser, FALSE, NULL);
+                return;
+        }
+
+        if (g_error_matches (error, G_IO_ERROR, G_IO_ERROR_CANCELLED)) {
+                sync_all_checkmarks (chooser, FALSE,
+                                     _("Canceled additional downloads"));
+                return;
+        }
+
+        sync_all_checkmarks (chooser, FALSE,
+                             _("Unable to download additional language packs"));
+        g_warning ("Installation failed: %s", error->message);
+}
+
 static void
 set_locale_id (CcLanguageChooser *chooser,
                const gchar       *new_locale_id)
@@ -407,9 +648,24 @@ set_locale_id (CcLanguageChooser *chooser,
         g_free (priv->language);
         priv->language = g_strdup (new_locale_id);
 
-        sync_all_checkmarks (chooser);
+        sync_all_checkmarks (chooser, TRUE, NULL);
+
+        g_autofree gchar *locale_for_install = NULL;
+        char *suffix = strchr (new_locale_id, '.');
+        if (suffix)
+                locale_for_install = g_strndup (new_locale_id, suffix - new_locale_id);
+        else
+                locale_for_install = g_strdup (new_locale_id);
 
-        g_object_notify_by_pspec (G_OBJECT (chooser), obj_props[PROP_LANGUAGE]);
+        g_cancellable_cancel (priv->install_cancellable);
+        g_clear_object (&priv->install_cancellable);
+        priv->install_cancellable = g_cancellable_new ();
+
+        g_autoptr (GTask) task = g_task_new (chooser, priv->install_cancellable,
+                                             installation_task_done, chooser);
+        g_task_set_task_data (task, g_steal_pointer (&locale_for_install), g_free);
+        g_task_set_return_on_cancel (task, TRUE);
+        g_task_run_in_thread (task, installation_task_async);
 }
 
 static gboolean
@@ -431,6 +687,8 @@ row_activated (GtkListBox        *box,
         GtkWidget *child;
         LanguageWidget *widget;
 
+        g_cancellable_cancel (priv->install_cancellable);
+
         if (row == NULL)
                 return;
 
@@ -477,7 +735,7 @@ cc_language_chooser_constructed (GObject *object)
         if (priv->language == NULL)
                 priv->language = cc_common_language_get_current_language ();
 
-        sync_all_checkmarks (chooser);
+        sync_all_checkmarks (chooser, FALSE, NULL);
 }
 
 static void
@@ -559,6 +817,10 @@ cc_language_chooser_class_init (CcLanguageChooserClass *klass)
                 g_param_spec_string ("showing-extra", "", "", "",
                                      G_PARAM_READABLE | G_PARAM_STATIC_STRINGS);
 
+        obj_props[PROP_BUSY] =
+                g_param_spec_boolean ("busy", "", "", FALSE,
+                                      G_PARAM_READABLE | G_PARAM_STATIC_STRINGS);
+
         g_object_class_install_properties (object_class, PROP_LAST, obj_props);
 }
 
@@ -595,3 +857,10 @@ cc_language_chooser_get_showing_extra (CcLanguageChooser *chooser)
         CcLanguageChooserPrivate *priv = cc_language_chooser_get_instance_private (chooser);
         return priv->showing_extra;
 }
+
+gboolean
+cc_language_chooser_get_busy (CcLanguageChooser *chooser)
+{
+        CcLanguageChooserPrivate *priv = cc_language_chooser_get_instance_private (chooser);
+        return priv->busy;
+}
diff --git a/gnome-initial-setup/pages/language/cc-language-chooser.h b/gnome-initial-setup/pages/language/cc-language-chooser.h
index 749af78..93afaf1 100644
--- a/gnome-initial-setup/pages/language/cc-language-chooser.h
+++ b/gnome-initial-setup/pages/language/cc-language-chooser.h
@@ -57,6 +57,7 @@ const gchar * cc_language_chooser_get_language (CcLanguageChooser *chooser);
 void          cc_language_chooser_set_language (CcLanguageChooser *chooser,
                                                 const gchar        *language);
 gboolean      cc_language_chooser_get_showing_extra (CcLanguageChooser *chooser);
+gboolean      cc_language_chooser_get_busy (CcLanguageChooser *chooser);
 
 G_END_DECLS
 
diff --git a/gnome-initial-setup/pages/language/gis-language-page.c b/gnome-initial-setup/pages/language/gis-language-page.c
index fe5fed9..1135a61 100644
--- a/gnome-initial-setup/pages/language/gis-language-page.c
+++ b/gnome-initial-setup/pages/language/gis-language-page.c
@@ -56,6 +56,16 @@ typedef struct _GisLanguagePagePrivate GisLanguagePagePrivate;
 
 G_DEFINE_TYPE_WITH_PRIVATE (GisLanguagePage, gis_language_page, GIS_TYPE_PAGE);
 
+static void
+language_chooser_busy_changed (GisLanguagePage *page)
+{
+  GisLanguagePagePrivate *priv = gis_language_page_get_instance_private (page);
+
+  gboolean complete = !cc_language_chooser_get_busy (
+    CC_LANGUAGE_CHOOSER (priv->language_chooser));
+  gis_page_set_complete (GIS_PAGE (page), complete);
+}
+
 static void
 set_localed_locale (GisLanguagePage *self)
 {
@@ -251,7 +261,15 @@ gis_language_page_constructed (GObject *object)
       g_object_unref (bus);
     }
 
-  gis_page_set_complete (GIS_PAGE (page), TRUE);
+  /* Block the user from advancing until the language chooser has completed
+   * updating the locale.
+   */
+  gis_page_set_complete (GIS_PAGE (page),
+                         !cc_language_chooser_get_busy (CC_LANGUAGE_CHOOSER(priv->language_chooser)));
+
+  g_signal_connect_object (priv->language_chooser, "notify::busy",
+                           G_CALLBACK (language_chooser_busy_changed), page,
+                           G_CONNECT_SWAPPED);
   gtk_widget_set_visible (GTK_WIDGET (page), TRUE);
 }
 
diff --git a/gnome-initial-setup/pages/software/gis-software-page.c b/gnome-initial-setup/pages/software/gis-software-page.c
index 0e2278f..6892365 100644
--- a/gnome-initial-setup/pages/software/gis-software-page.c
+++ b/gnome-initial-setup/pages/software/gis-software-page.c
@@ -96,7 +96,7 @@ gis_software_page_apply (GisPage      *gis_page,
       program = find_fedora_third_party ();
       if (program)
         {
-          gis_pkexec (program, "enable", "root", &error);
+          gis_pkexec (program, "enable", "root", cancellable, &error);
           if (error && !g_error_matches (error, G_IO_ERROR, G_IO_ERROR_CANCELLED))
             g_warning ("%s failed: %s", program, error->message);
         }
